### PR TITLE
Clone polishing

### PR DESCRIPTION
**Pull request by georgeliao**
[MULTI-1558](https://warthogs.atlassian.net/browse/MULTI-1558)

[MULTI-1558]: https://warthogs.atlassian.net/browse/MULTI-1558?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/daemon/daemon.cpp [hunk 1]
```diff
@@ -2706,29 +2705,32 @@ try
...
-                             "Please stop instance " + source_name + " before you clone it."});
+                grpc::Status{grpc::FAILED_PRECONDITION, "Multipass can only clone stopped instances."});
         }
 
-        const std::string destination_name = generate_destination_instance_name_for_clone(*request);
-        auto rollback_clean_up_all_resource_of_dest_instance =
-            sg::make_scope_guard([this, destination_name]() noexcept -> void {
+        const std::string destination_name = dest_name_for_clone(*request);
+        if (auto dest_vm_status = validate_dest_name(destination_name); !dest_vm_status.ok())
+            return status_promise->set_value(std::move(dest_vm_status));
```
Comment (georgeliao) [inline src/daemon/daemon.cpp:2726]: Refactor `create_vm` as well based this?

**Review by ricab — COMMENTED** [review]
Looks great, good job! Thanks for all the effort that you put into these improvements. 

I have only a few small details on this side.

File: src/client/cli/cmd/restore.cpp [hunk 2]
```diff
@@ -87,7 +87,7 @@ QString cmd::Restore::short_help() const
 
 QString cmd::Restore::description() const
 {
-    return QStringLiteral("Restore an instance to the state of a previously taken snapshot.");
+    return QStringLiteral("Restore an (stopped) instance to the state of a previously taken snapshot.");
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/client/cli/cmd/restore.cpp]: Done

File: src/client/cli/cmd/snapshot.cpp [hunk 3]
```diff
@@ -64,7 +64,8 @@ QString cmd::Snapshot::short_help() const
 
 QString cmd::Snapshot::description() const
 {
-    return QStringLiteral("Take a snapshot of an instance that can later be restored to recover the current state.");
+    return QStringLiteral(
+        "Take a snapshot of an (stopped) instance that can later be restored to recover the current state.");
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/client/cli/cmd/snapshot.cpp]: Done

File: src/platform/backends/shared/base_virtual_machine_factory.cpp [hunk 4]
```diff
@@ -83,27 +84,61 @@ void mp::BaseVirtualMachineFactory::prepare_interface(NetworkInterface& net,
...
+                                                                      const VMSpecs& dest_spec,
+                                                                      const std::string& src_name,
+                                                                      const std::string& dest_name,
+                                                                      const VMImage& dest_image,
+                                                                      const multipass::SSHKeyProvider& key_provider,
+                                                                      VMStatusMonitor& monitor)
 {
-    if (fs::exists(source_instance_dir_path) && fs::is_directory(source_instance_dir_path))
+    const std::filesystem::path source_instance_data_directory{get_instance_directory(src_name).toStdString()};
+    const std::filesystem::path dest_instance_data_directory{get_instance_directory(dest_name).toStdString()};
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/shared/base_virtual_machine_factory.cpp]: Done

File: src/utils/yaml_node_utils.cpp [hunk 5]
```diff
@@ -116,6 +116,9 @@ YAML::Node mp::utils::make_cloud_init_network_config(const std::string& default_
                                                      const std::vector<mp::NetworkInterface>& extra_interfaces,
                                                      const std::string& file_content)
 {
+    // It is changed to generate new format network-config file, meaning always be present, have default interface
+    // field and dhcp-identifier: mac on every network interface. It also remains backward compatible to the absent
+    // network-config/default interface (the old format) cases.
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/utils/yaml_node_utils.cpp]: Done

File: src/utils/yaml_node_utils.cpp [hunk 6]
```diff
@@ -137,6 +140,7 @@ YAML::Node mp::utils::add_extra_interface_to_network_config(const std::string& d
                                                             const NetworkInterface& extra_interface,
                                                             const std::string& network_config_file_content)
 {
+    // Same as the make_cloud_init_network_config function in terms of format compatibility
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/utils/yaml_node_utils.cpp]: Done.

File: src/platform/backends/shared/base_virtual_machine_factory.h [hunk 7]
```diff
@@ -103,12 +95,17 @@ class BaseVirtualMachineFactory : public VirtualMachineFactory
     virtual void remove_resources_for_impl(const std::string& name) = 0;
 
 private:
+    virtual VirtualMachine::UPtr clone_vm_impl(const std::string& source_vm_name,
+                                               const multipass::VMSpecs& src_vm_specs,
+                                               const VirtualMachineDescription& desc,
+                                               VMStatusMonitor& monitor,
+                                               const SSHKeyProvider& key_provider);
+    static void copy_instance_dir_with_essential_files(const fs::path& source_instance_dir_path,
```
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/shared/base_virtual_machine_factory.h:103]: OK, out of scope of this PR in any case so leaving it alone.

**Review by ricab — DISMISSED** [review]
Great, thanks again!

**Review by ricab — APPROVED** [review]

**Comment by ricab** [comment]
@georgeliao It looks like something is amiss on the Windows side:

```
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] HyperVBackend.throws_on_failure_to_setup_default_network_adapter
[  FAILED  ] HyperVBackend.throws_on_failure_to_detect_switch_from_extra_interface
[  FAILED  ] HyperVBackend.throws_on_failure_to_add_extra_interface
```

I have seen that on 2 runs now.

**Comment by georgeliao** [comment]
> @georgeliao It looks like something is amiss on the Windows side:
> 
> [  FAILED  ] 3 tests, listed below:
> [  FAILED  ] HyperVBackend.throws_on_failure_to_setup_default_network_adapter
> [  FAILED  ] HyperVBackend.throws_on_failure_to_detect_switch_from_extra_interface
> [  FAILED  ] HyperVBackend.throws_on_failure_to_add_extra_interface
> I have seen that on 2 runs now.

Thanks for noticing that, I already know which commit caused that and I will have fix soon.

**Comment by georgeliao** [comment]
@ricab 

> [  FAILED  ] 3 tests, listed below:
> [  FAILED  ] HyperVBackend.throws_on_failure_to_setup_default_network_adapter
> [  FAILED  ] HyperVBackend.throws_on_failure_to_detect_switch_from_extra_interface
> [  FAILED  ] HyperVBackend.throws_on_failure_to_add_extra_interface
> I have seen that on 2 runs now.

I found the root cause of the failure. Basically, wrapping the constructor's data member list into private constructor changes the behavior when the constructor body throws. That is the destructor of the class (`HyperVVirtualMachine` in this case) gets called, and it was not called in the inlined case. I think the complier checks whether the constructor actually finishes or not to decides whether the desctructor should be called. In this case of private constructor wrapper, at least the internal constructor is finished. As a result, it decides to call destructor. 
Feel free to play with the [code snippet](https://godbolt.org/z/9zW3xMKPW) to understand this.

**Comment by ricab** [comment]
The merge queue will fail on mac/win, but I'd still like to verify Linux. Unless it starts taking long for any reason, but worth a shot...

**Pull request merged**